pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartialUpsertHandler.java

```diff
@@ -41,16 +44,33 @@ public class PartialUpsertHandler {
   private final PartialUpsertMerger _defaultPartialUpsertMerger;
   private final List<String> _comparisonColumns;
   private final List<String> _primaryKeyColumns;
+  private final PartialUpsertRowMergeEvaluator _rowMerger;
+  private LazyRow _reusePreviousLazyRow;
+  private LazyRow _reuseNewLazyRow;
+  private Map<String, Object> _reuseRowMergerResult;
 
-  public PartialUpsertHandler(Schema schema, Map<String, UpsertConfig.Strategy> partialUpsertStrategies,
-      UpsertConfig.Strategy defaultPartialUpsertStrategy, List<String> comparisonColumns) {
-    _defaultPartialUpsertMerger = PartialUpsertMergerFactory.getMerger(defaultPartialUpsertStrategy);
+  public PartialUpsertHandler(Schema schema, UpsertConfig upsertConfig, List<String> comparisonColumns) {
+    _defaultPartialUpsertMerger = PartialUpsertMergerFactory.getMerger(upsertConfig.getDefaultPartialUpsertStrategy());
     _comparisonColumns = comparisonColumns;
     _primaryKeyColumns = schema.getPrimaryKeyColumns();
 
-    for (Map.Entry<String, UpsertConfig.Strategy> entry : partialUpsertStrategies.entrySet()) {
+    for (Map.Entry<String, UpsertConfig.Strategy> entry : upsertConfig.getPartialUpsertStrategies().entrySet()) {
       _column2Mergers.put(entry.getKey(), PartialUpsertMergerFactory.getMerger(entry.getValue()));
     }
+
+    String rowMergerCustomImplementation = upsertConfig.getRowMergerCustomImplementation();
+    if (rowMergerCustomImplementation != null && !rowMergerCustomImplementation.equals("")) {
+      try {
+        _rowMerger = PartialUpsertRowMergeEvaluatorFactory.getInstance(rowMergerCustomImplementation);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      _reusePreviousLazyRow = new LazyRow();
+      _reuseNewLazyRow = new LazyRow();
+      _reuseRowMergerResult = new HashMap<>();
```

Comment:
concurrent hashmap?

Reply:
Should not be needed since this is per partition.